src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs

```diff
@@ -204,7 +204,11 @@ internal HttpRetryHandler(IEnvironmentVariableReader environmentVariableReader)
                             requestUri,
                             bodyStopwatch.ElapsedMilliseconds));
 
-                        if ((int)response.StatusCode >= 500)
+                        int statusCode = (int)response.StatusCode;
+                        // 5xx == server side failure
+                        // 408 == request timeout
+                        // 429 == too many requests
+                        if (statusCode >= 500 || statusCode == 408 || statusCode == 429)
```

Comment:
I remember when working on enhanced retry that not all code paths use it because they tell this code to not retry and then it has its own retry.  The backoff strategy in enhanced retry should keep us from causing more throttling, but have you looked at more callers to this method to make sure they won't retry without a backoff?  I just don't want to make the throttling worse...

Reply:
+1 on above one. Sudden burst of 64 connections (we increased recently from 2) repeatedly calling without giving any change to recover.

Reply:
> Sudden burst of 64 connections (we increased recently from 2) 

That's only in VS. 

We were using max on the commandline already.

Reply:
> The backoff strategy in enhanced retry should keep us from causing more throttling, but have you looked at more callers to this method to make sure they won't retry without a backoff? 

I'm aware that some "resources" do their own retry logic, which means that the enhanced retry can't do its exponential backoff.  However, the implementation in this scenario is that it will do a fixed duration wait, rather than exponential backoff. Given the default retry count is 3, I don't expect the difference between fixed vs exponential backoff will be very great.

> I just don't want to make the throttling worse...

Me either, but code that does its own retry loop isn't what I'm concerned about. I'm more concerned about whether a 1s (or less) retry is long enough if the Retry-After header requests to wait for 60 seconds. So, it might make the server load problem worse, but on the other hand failing restore because a retryable scenario wasn't retried also has bad customer experience. Which one is worse is a matter of debate, and would probably be very customer specific, rather than a generalization that applies to most customers. Even for a specific scenario (like the dotnet org who are experiencing it on Azure Artifacts), I don't think we have enough data to analyzer whether it would be beneficial or harmful.

I am planning on working on 1. retry 429's, 2. use Retry-After header and 3. consolidate all retries, so exponential backoff (and Retry-After) only needs to be implemented in a single place. My question is do we wait for everything to be implemented before we take any change, or if some of these changes are independently beneficial, do we take the benefits as they're ready?

Reply:
I'm personally not as concerned about doubling on the throttling issues as I don't expect the backoff is going to be long enough/too long. 

If we're concerned, we should put this change behind an environment variable as the change gets adopted within the .NET SDK repos.